src/Features/Core/Portable/UseCompoundAssignment/AbstractUseCompoundAssignmentCodeFixProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.UseCompoundAssignment
+{
+    internal abstract class AbstractUseCompoundAssignmentCodeFixProvider<
+        TSyntaxKind, TAssignmentSyntax, TExpressionSyntax> 
+        : SyntaxEditorBasedCodeFixProvider
+        where TSyntaxKind : struct
+        where TAssignmentSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(IDEDiagnosticIds.UseCompoundAssignmentDiagnosticId);
+
+        // See comments in the analyzer for what these maps are for.
+
+        private readonly ImmutableDictionary<TSyntaxKind, TSyntaxKind> _binaryToAssignmentMap;
+        private readonly ImmutableDictionary<TSyntaxKind, TSyntaxKind> _assignmentToTokenMap;
+
+        protected AbstractUseCompoundAssignmentCodeFixProvider(
+            ImmutableArray<(TSyntaxKind exprKind, TSyntaxKind assignmentKind, TSyntaxKind tokenKind)> kinds)
+        {
+            Utilities.GenerateMaps(kinds, out _binaryToAssignmentMap, out _assignmentToTokenMap);
+        }
+
+        protected abstract TSyntaxKind GetSyntaxKind(int rawKind);
+        protected abstract SyntaxToken Token(TSyntaxKind kind);
+        protected abstract TAssignmentSyntax Assignment(
+            TSyntaxKind assignmentOpKind, TExpressionSyntax left, SyntaxToken syntaxToken, TExpressionSyntax right);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var diagnostic = context.Diagnostics[0];
+            
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(document, diagnostic, c)),
+                context.Diagnostics);
+
+            return Task.CompletedTask;
+        }
+
+        protected override Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var assignment = diagnostic.AdditionalLocations[0].FindNode(getInnermostNodeForTie: true, cancellationToken);
```

Comment:
this assumes the diagnostic you got is the diagnostic you produced. anyone can produce diagnostic with same id.

Reply:
the IDE assumption is that anyone producing a diagnostic must abide by the contract of that diagnostic.  i.e. the diagnostic has expectations about what hte span is, what the additional locatoins are, etc. etc.  tehre is no validation done that these match.

Reply:
Note: this is hte pattern across 100% of IDE analyzers/fixers.  There is no validation done of diagnostics.  Assumption is that these our our diagnostics (with our 'IDE' prefix) and are not for others to produce.

Reply:
either it is existing pattern or not. I am stating that it can cause issue if any analyzer produce IDE diagnostic IDs. there is no one prevent it or contact says you can't do it.